setup.py

```diff
@@ -19,7 +19,7 @@
         "pangeo-forge-recipes>=0.9.2",
         "escapism",
         "traitlets",
-        "apache-beam[gcp]",
+        "apache-beam[gcp]==2.47.0",
```

Comment:
You're probably already tracking this (and have made above edit for convenience), but ultimately I think we'll want to specify this version pin in either:

- https://github.com/ranchodeluxe/gpcp-from-gcs-feedstock/blob/main/feedstock/requirements.txt
- or possibly a Flink-specific set of optional dependencies?

Reply:
yeah, for convenience right now

Reply:
Cool cool, figured 😎

Reply:
@cisaacstern: my hunch was your suggestion above wouldn't work based on what I've been seeing the last week. And it broke the [passing tests](https://github.com/pangeo-forge/pangeo-forge-runner/actions/workflows/flink.yaml?query=branch%3Agcorradini%2Funpin-beam-fix-integration-test)

The `apache-beam` version designated by `pangeo-forge-runner` will determine which job-server jar is download and uploaded to the flink server. Note below it's uploading `2.51.0` even though I have my recipe pinned to `apache-beam==2.47.0`:

<img width="1376" alt="Screen Shot 2023-10-25 at 7 12 04 AM" src="https://github.com/pangeo-forge/pangeo-forge-runner/assets/208859/c33a1f7f-8afd-4fc5-880e-8064db66370d">

So there's currently a gross tight coupling between producer and consumer versions of beam versions  that has be coordinated in the following places:
producer: pangeo-forge-runner `setup.py`
consumer: the [container_image](https://github.com/pangeo-forge/pangeo-forge-runner/pull/114/files#diff-f60f5f1833da4ce45d7eeb715ebc93ac07540ccea9e1cb238bf4d661dd8dcf30R37)

Let me recover the actual logs from the Flink run to determine what the issue is but I've seen it talk about incompatible version before and that's my guess what's happening here

Maybe @yuvipanda [suggestion here](https://github.com/pangeo-forge/pangeo-forge-runner/issues/111#issuecomment-1774819567) is that we try to run Flink without uploading the job-server jar? Maybe I can investigate that

Reply:
Nevermind, this repository is archived: https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/tree/master/examples/beam/with_job_server

Reply:
@ranchodeluxe so on GCP at least, I think beam knows how to do this coupling between producer and consumer versions correctly so it's a non-issue. Makes sense it'll need to be somehow handled explicitly here on flink.

Reply:
Yes don't try to do what I did, it ends in frustration and the desire to erase the JVM from historical existence.

Reply:
Thanks for all your awesome work here @ranchodeluxe.

For expedience to get this all merged, maybe we remove beam from the top-level dependencies and do:

```python
setup(
    ...,
    extras_require={
        "dataflow": ["apache-beam[gcp]"],
        "flink": ["apache-beam==2.47.0"],
    },
)
```
?

Any thoughts on that approach @yuvipanda or @ranchodeluxe ?

Reply:
> For expedience to get this all merged, maybe we remove beam from the top-level dependencies and do:

yes, @cisaacstern this makes sense but I'd do `"flink": ["apache-beam>=2.47.0"]` to convey that it should work from that version forward and then I can have the integration tests use a build matrix for those different versions of `apache-beam`

Try things out now for different versions of `apache-beam` 🤞

Reply:
> Note below it's uploading 2.51.0 even though I have my recipe pinned to apache-beam==2.47.0

Hmm actually now that I think about it this is probably just because the integration test is not installing the recipe's requirements.txt in the producer/client/deployer environment? We have a long-standing issue related to that: https://github.com/pangeo-forge/pangeo-forge-runner/issues/27, but currently it's just up to the deployer to make sure that happens on the client side. I've hacked in a solution for that [in the GitHub Action here](https://github.com/pangeo-forge/deploy-recipe-action/blob/f0ec71a5adc128abe5e1df733ea7d4434b5c0560/action/deploy_recipe.py#L88-L94).

Reply:
Or I should say, in the GCP Dataflow context, the beam version for the worker container is inferred from the client environment, but maybe that assumption doesn't hold for Flink.

Reply:
> the beam version for the worker container is inferred from the client 

Yeah, this is what I'd like to happen.  And it's good for me to review between the runners. 

Currently `Bake.container_image` defaults to `""` which our doc strings say should force beam to figure out which worker SDK image to use. This might work for GCP DataFlow but for Flink [we override our `container` definition (among other things) for the pod to use that specific image](https://github.com/pangeo-forge/pangeo-forge-runner/blob/main/pangeo_forge_runner/bakery/flink.py#L177).

So using the default `""` for Fink breaks and we are forced to supply one (this is a bug I have listed that needs a validation fix in my queue of things to do)

> is not installing the recipe's requirements.txt in the producer/client/deployer environment?

This is curious and you are right that the apache-beam client/pipeline stages everything from the recipe's `requirement.txt` except the `apache-beam` packages even if they exist in there 🤔 I assume it's doing this for a smart reason b/c the job-server jar already knows what versions it's dealing with but this is where my understanding of the architecture breaks down